### PR TITLE
Eliminate binary.Read in instruction (un)marshaler

### PR DESCRIPTION
**Pull request by ti-mo**
This still needs to be fuzzed, but the gains are nice so far:

```
name                    old time/op    new time/op    delta
Read64bitImmediate-16      479ns ±12%      74ns ± 1%  -84.62%  (p=0.008 n=5+5)
Write64BitImmediate-16     417ns ±10%      63ns ± 2%  -84.97%  (p=0.008 n=5+5)

name                    old alloc/op   new alloc/op   delta
Read64bitImmediate-16      32.0B ± 0%      8.0B ± 0%  -75.00%  (p=0.008 n=5+5)
Write64BitImmediate-16     24.0B ± 0%      8.0B ± 0%  -66.67%  (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
Read64bitImmediate-16       4.00 ± 0%      1.00 ± 0%  -75.00%  (p=0.008 n=5+5)
Write64BitImmediate-16      3.00 ± 0%      1.00 ± 0%  -66.67%  (p=0.008 n=5+5)
```

The use of `internal.DiscardZeroes{}` was considered for checking for `64bit immediate has non-zero fields`, but ended up being a lot slower than just including those bits in the read. The fact that it aligns with a single u32 interpretation is quite convenient. :)

### CODE REVIEW

File: asm/instruction.go [hunk 1]
```diff
@@ -47,33 +47,46 @@ func (ins Instruction) Sym(name string) Instruction {
...
+
+	if !ins.OpCode.IsDWordLoad() {
 		return InstructionSize, nil
 	}
 
-	var bi2 bpfInstruction
-	if err := binary.Read(r, bo, &bi2); err != nil {
+	// Pull another instruction from the stream to retrieve the second
+	// half of the 64-bit immediate value.
+	data = make([]byte, InstructionSize)
```
Comment (brycekahle) [inline asm/instruction.go:76]: Do you need the second `make` here? You are going to fill the entire slice since you are using `io.ReadFull`.

Comment (ti-mo) [reply asm/instruction.go:76]: Thanks, not sure how I missed that.

File: asm/instruction.go [hunk 2]
```diff
@@ -97,26 +109,22 @@ func (ins Instruction) Marshal(w io.Writer, bo binary.ByteOrder) (uint64, error)
...
 	if !isDWordLoad {
 		return InstructionSize, nil
 	}
 
-	bpfi = bpfInstruction{
-		Constant: int32(ins.Constant >> 32),
-	}
-
-	if err := binary.Write(w, bo, &bpfi); err != nil {
+	data = make([]byte, InstructionSize)
```
Comment (lmb) [inline asm/instruction.go:125]: Idea: re-use the existing allocation or allocate 2*InstructionSize above?

File: asm/instruction.go [hunk 3]
```diff
@@ -527,13 +535,6 @@ func (iter *InstructionIterator) Next() bool {
...
 }
 
-type bpfInstruction struct {
-	OpCode    OpCode
-	Registers bpfRegisters
-	Offset    int16
-	Constant  int32
-}
-
 type bpfRegisters uint8
```
Comment (lmb) [inline asm/instruction.go:538]: Is this still used?

File: asm/instruction_test.go [hunk 4]
```diff
@@ -48,6 +60,21 @@ func TestWrite64bitImmediate(t *testing.T) {
...
+func BenchmarkWrite64BitImmediate(b *testing.B) {
+	insns := Instructions{
+		LoadImm(R0, math.MinInt32-1, DWord),
+	}
+
+	var buf bytes.Buffer
+	for i := 0; i < b.N; i++ {
+		buf.Reset()
+
+		if err := insns.Marshal(&buf, binary.LittleEndian); err != nil {
```
Comment (lmb) [inline asm/instruction_test.go:72]: Nit: why not benchmark Instruction.Marshal directly?

**Review by lmb — APPROVED** [review]
Nice!

File: asm/instruction.go (hunk 2 above)
Comment (ti-mo) [reply asm/instruction.go:125]: Good catch, I think zeroing out the first 4 bytes is the way to go.

File: asm/instruction_test.go (hunk 4 above)
Comment (ti-mo) [reply asm/instruction_test.go]: Not sure either, fixed.

**Pull request merged**